std/algorithm/iteration.d

```diff
@@ -902,67 +913,100 @@ template each(alias pred = "a")
             size_t i = 0;
             while (!r.empty)
             {
-                cast(void) binaryFun!BinaryArgs(i, r.front);
+                static if (!is(typeof(binaryFun!BinaryArgs(i, r.front)) ==
+                        Flag!"each"))
+                    cast(void) binaryFun!BinaryArgs(i, r.front);
```

Comment:
Depends on whether you want `each` to only accept non-strongly-pure functions or not. My first thought is that it's fine because why would you use `each` if you don't want to do something non-pure? However, if you want `each` to be as close in functionality to `foreach` as possible, then strongly-pure functions should be allowed as it is perfectly acceptable to have a "pure" foreach loop:

```
foreach (i; [0, 1, 2])
{
    //This loop has no side-effects
}
```